# autocfg.py

## This example notebook shows how to use the autocfg.py file in saspy to automatically generate a working saspy config file for connecting to a local SAS install on a Windows system.
## autocfg.py can be run as a batch script, or from within python.

## Even if this works for you, this is no excuse for not reading the install/config doc :)
## https://sassoftware.github.io/saspy/install.html

### There are three optional parameters you can provide. 
* cfgfile - this is the path to the config file that will be created. By default it will be sascfg_personal.py in the saspy install directory
* SASHome - this is the path to your SASHome directory; the SAS installation directory. It defaults to C:\Program Files\SASHome which is the SAS default install path
* Java - this is the path to your java.exe on your system. It defaults to java which is usually found in your path when you install java on Windwos

## These parameters can be specified positionally or by keyword interactivly. For batch, they are positional only, but you can specify None to not specify some. See the examples for that below.



### Here is the signature for the function. The interactive use case follows

In [1]:
#def main(cfgfile: str = None, SASHome: str = None, java: str = None):

## You can create a Windows local sascfg_personal.py file using autocfg.py from within Python
## Here is the simplest case, for a freshly installed saspy. It will create a config file that works imediatly; assuming I have SAS and Java installed in default locations

In [2]:
from saspy import autocfg
autocfg.main()

Generated configurations file: C:\ProgramData\Anaconda3\lib\site-packages\saspy\sascfg_personal.py



In [3]:
import saspy
sas = saspy.SASsession()
sas

Using SAS Config named: autogen_winlocal
SAS Connection established. Subprocess id is 11220



Access Method         = IOM
SAS Config name       = autogen_winlocal
WORK Path             = C:\Users\sastpw\AppData\Local\Temp\SAS Temporary Files\_TD15016_d10a626_\Prc2\
SAS Version           = 9.04.01M5P09132017
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = windows-1252
SAS process Pid value = 15016


In [4]:
fd = open('C:\ProgramData\Anaconda3\lib\site-packages\saspy\sascfg_personal.py')
print(fd.read())
fd.close()

SAS_config_names=["autogen_winlocal"]

SAS_config_options = {
	"lock_down": False,
	"verbose"  : True
	}

cpW  =  "C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.security.sspi.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.svc.connection.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\log4j.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.core.jar"
cpW += ";C:\\ProgramData\\Anaconda3\\lib\\site-packages\\saspy\\java\\saspyiom.jar"

autogen_winlocal = {
	"java"      : "java",
	"encoding"  : "windows-1252",
	"classpath" : cpW
	}

import os
os.environ["PATH"] += ";C:\\Program Files\\SASHome\\SASFoundation\\9.4\\core\\sasext"


## You can create a different sascfg file that you can refer to in SASsession()
## Also, lets specify Java ourselves to see that it is used in the config file

In [5]:
from saspy import autocfg
autocfg.main('my_cfg.py', java=r'C:\ProgramData\Oracle\Java\javapath\java.exe')

Generated configurations file: my_cfg.py


In [6]:
import saspy
sas = saspy.SASsession(cfgfile='my_cfg.py')
sas

Using SAS Config named: autogen_winlocal
SAS Connection established. Subprocess id is 8248



Access Method         = IOM
SAS Config name       = autogen_winlocal
WORK Path             = C:\Users\sastpw\AppData\Local\Temp\SAS Temporary Files\_TD4452_d10a626_\Prc2\
SAS Version           = 9.04.01M5P09132017
SASPy Version         = 2.4.1
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = windows-1252
SAS process Pid value = 4452


In [7]:
fd = open('my_cfg.py')
print(fd.read())
fd.close()

SAS_config_names=["autogen_winlocal"]

SAS_config_options = {
	"lock_down": False,
	"verbose"  : True
	}

cpW  =  "C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.svc.connection.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.security.sspi.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.core.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\log4j.jar"
cpW += ";C:\\ProgramData\\Anaconda3\\lib\\site-packages\\saspy\\java\\saspyiom.jar"

autogen_winlocal = {
	"java"      : "C:\\ProgramData\\Oracle\\Java\\javapath\\java.exe",
	"encoding"  : "windows-1252",
	"classpath" : cpW
	}

import os
os.environ["PATH"] += ";C:\\Program Files\\SASHome\\SASFoundation\\9.4\\core\\sasext"


## Now lets see how to run this as a Batch script

### First, run the default, right after a fresh install of saspy 

```
C:\tom1>python \ProgramData\Anaconda3\Lib\site-packages\saspy\autocfg.py
Generated configurations file: C:\ProgramData\Anaconda3\lib\site-packages\saspy\sascfg_personal.py


C:\tom1>type C:\ProgramData\Anaconda3\lib\site-packages\saspy\sascfg_personal.py
SAS_config_names=["autogen_winlocal"]

SAS_config_options = {
        "lock_down": False,
        "verbose"  : True
        }

cpW  =  "C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\log4j.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.core.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.svc.connection.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.security.sspi.jar"
cpW += ";C:\\ProgramData\\Anaconda3\\lib\\site-packages\\saspy\\java\\saspyiom.jar"

autogen_winlocal = {
        "java"      : "java",
        "encoding"  : "windows-1252",
        "classpath" : cpW
        }

import os
os.environ["PATH"] += ";C:\\Program Files\\SASHome\\SASFoundation\\9.4\\core\\sasext"

C:\tom1>python
Python 3.6.0 |Anaconda 4.3.0 (64-bit)| (default, Dec 23 2016, 11:57:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import saspy
>>> sas = saspy.SASsession()
Using SAS Config named: autogen_winlocal
SAS Connection established. Subprocess id is 27532

>>> sas
Access Method         = IOM
SAS Config name       = autogen_winlocal
WORK Path             = C:\Users\sastpw\AppData\Local\Temp\SAS Temporary Files\_TD27852_d10a626_\Prc2\
SAS Version           = 9.04.01M5P09132017
SASPy Version         = 2.4.0
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = windows-1252
SAS process Pid value = 27852

>>>
```

### Now lets create a different sascfg file that you can refer to in SASsession()
### And, lets specify Java ourselves to see that it is used in the config file
* Note that Java is the 3rd positional parm. To skip SASHome, just specify None
* also, use double quotes to enclose paths, especially when they have spaces in them like "C:\Program Files\SASHome"

<pre>
C:\tom1>python \ProgramData\Anaconda3\Lib\site-packages\saspy\autocfg.py batch_cfg.py None "C:\ProgramData\Oracle\Java\javapath\java.exe"
Generated configurations file: batch_cfg.py


C:\tom1>type batch_cfg.py
SAS_config_names=["autogen_winlocal"]

SAS_config_options = {
        "lock_down": False,
        "verbose"  : True
        }

cpW  =  "C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.core.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.svc.connection.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.security.sspi.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\log4j.jar"
cpW += ";C:\\ProgramData\\Anaconda3\\lib\\site-packages\\saspy\\java\\saspyiom.jar"

autogen_winlocal = {
        "java"      : "C:\\ProgramData\\Oracle\\Java\\javapath\\java.exe",
        "encoding"  : "windows-1252",
        "classpath" : cpW
        }

import os
os.environ["PATH"] += ";C:\\Program Files\\SASHome\\SASFoundation\\9.4\\core\\sasext"

C:\tom1>python
Python 3.6.0 |Anaconda 4.3.0 (64-bit)| (default, Dec 23 2016, 11:57:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import saspy
>>> sas = saspy.SASsession(cfgfile='batch_cfg.py')
Using SAS Config named: autogen_winlocal
SAS Connection established. Subprocess id is 14692

>>> sas
Access Method         = IOM
SAS Config name       = autogen_winlocal
WORK Path             = C:\Users\sastpw\AppData\Local\Temp\SAS Temporary Files\_TD28096_d10a626_\Prc2\
SAS Version           = 9.04.01M5P09132017
SASPy Version         = 2.4.0
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = windows-1252
SAS process Pid value = 28096

>>>
</pre>

### One last case showing all three parameters. 

<pre>
C:\tom1>python \ProgramData\Anaconda3\Lib\site-packages\saspy\autocfg.py \ProgramData\Anaconda3\Lib\site-packages\saspy\all_parms.py "C:\Program Files\SASHome" java.exe
Generated configurations file: \ProgramData\Anaconda3\Lib\site-packages\saspy\all_parms.py


C:\tom1>type  \ProgramData\Anaconda3\Lib\site-packages\saspy\all_parms.py
SAS_config_names=["autogen_winlocal"]

SAS_config_options = {
        "lock_down": False,
        "verbose"  : True
        }

cpW  =  "C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\log4j.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.core.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.security.sspi.jar"
cpW += ";C:\\Program Files\\SASHome\\SASDeploymentManager\\9.4\\products\\deploywiz__94508__prt__xx__sp0__1\\deploywiz\\sas.svc.connection.jar"
cpW += ";C:\\ProgramData\\Anaconda3\\lib\\site-packages\\saspy\\java\\saspyiom.jar"

autogen_winlocal = {
        "java"      : "java.exe",
        "encoding"  : "windows-1252",
        "classpath" : cpW
        }

import os
os.environ["PATH"] += ";C:\\Program Files\\SASHome\\SASFoundation\\9.4\\core\\sasext"

C:\tom1>python
Python 3.6.0 |Anaconda 4.3.0 (64-bit)| (default, Dec 23 2016, 11:57:41) [MSC v.1900 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import saspy
>>> sas = saspy.SASsession(cfgfile='c:\\ProgramData\\Anaconda3\\Lib\\site-packages\\saspy\\all_parms.py')
Using SAS Config named: autogen_winlocal
SAS Connection established. Subprocess id is 19576

>>> sas
Access Method         = IOM
SAS Config name       = autogen_winlocal
WORK Path             = C:\Users\sastpw\AppData\Local\Temp\SAS Temporary Files\_TD28236_d10a626_\Prc2\
SAS Version           = 9.04.01M5P09132017
SASPy Version         = 2.4.0
Teach me SAS          = False
Batch                 = False
Results               = Pandas
SAS Session Encoding  = wlatin1
Python Encoding value = windows-1252
SAS process Pid value = 28236

>>>
</pre>